server/metrics/metrics.go

```diff
@@ -99,6 +99,14 @@ type metrics struct {
 	storeTimesHistograms *prometheus.HistogramVec
 }
 
+// NilMetrics returns the Metrics interface with a concrete, but nil value. Since all methods on
+// the metrics type are safe to call with a nil receiver, this gives the caller a "no-op" version
+// of the metrics interface to use when metrics are disabled.
+func NilMetrics() Metrics {
+	var m *metrics
+	return m
+}
+
```

Comment:
This is an important part of the change.

Reply:
I don't fully understand why we need that trick, In theory the attribute should be already a nil interface with the concrete type defined by the attribute. I guess you already have checked that. But I don't understand the why behind.

Reply:
Until we set `p.metricsService := NewMetrics(...)`, the interface is a `(value, type)` of `(nil, nil)`. So none of the `m != nil` safeguards apply since there's no concrete type. 

The `NilMetrics` constructor returns an interface tuple of `(nil, Metrics)` which -- combined with the `m != nil` implementation allows safe calling despite the value not being initialized.

Another approach -- less magic -- is just to instantiate with `NewMetrics()` and simply decide whether or not to expose the port via the config. So always collect metrics, unconditionally.

Thoughts?

Reply:
@wiggin77, the `NilMetrics` is something I'm thinking is unnecessarily complexity. I'm inclined to just always return the metrics, but only conditionally expose the port. Thoughts?

Reply:
Agreed.

Reply:
Sounds good: done!